R/data.R

```diff
@@ -123,69 +123,96 @@ any_na_elements <- function(data) {
 }
 
 
+draws_to_csv <- function(draws, sampler_diagnostics = NULL) {
+  sampler_diagnostic_names <- c("accept_stat__", "stepsize__", "treedepth__", "n_leapfrog__", "divergent__", "energy__")
+
+  n <- posterior::niterations(draws)
+  n_chains <- posterior::nchains(draws)
+  if (is.null(sampler_diagnostics)) {
+    # create dummy sampler diagnostics due to CmdStan requirement for all columns in GQ
+    sampler_diagnostics <- rep(0, n * length(sampler_diagnostic_names) * n_chains)
+    dim(sampler_diagnostics) <- c(n, n_chains, length(sampler_diagnostic_names))
+    sampler_diagnostics <- posterior::as_draws_array(sampler_diagnostics)
+    posterior::variables(sampler_diagnostics) <- sampler_diagnostic_names
+  }
+
+  # the columns must be in order "lp__, sampler_diagnostics, parameters"
+  variables <- posterior::variables(draws)
+  # create a dummy lp__ column if it does not exist
+  if ("lp__" %in% variables) {
+    lp__ <- NULL
+  } else {
+    lp__ <- rep(0, n * n_chains)
+    dim(lp__) <- c(n, n_chains, 1)
+    lp__ <- posterior::as_draws_array(lp__)
+    posterior::variables(lp__) <- "lp__"
+  }
+  variables <- c("lp__", sampler_diagnostic_names, variables[!(variables %in% c("lp__", "lp_approx__"))])
+  draws <- posterior::subset_draws(
+    posterior::bind_draws(draws, sampler_diagnostics, lp__, along = "variable"),
+    variable = variables
+  )
+  chains <- posterior::chain_ids(draws)
+  paths <- generate_file_names(basename = "fittedParams", ids = chains)
+  paths <- file.path(tempdir(), paths)
+  chain <- 1
+  for (path in paths) {
+    write(
+      paste0("# num_samples = ", n, "\n", paste0(unrepair_variable_names(variables), collapse = ",")),
+      file = path,
+      append = FALSE
+    )
+    utils::write.table(
```

Comment:
Does it make sense to use `data.table::fwrite()` here to write the files instead of `utils::write.table()`? I'm not sure how big of a difference the speed is for writing, but just asking since we're already using `data.table::fread()`. We could also just use write.table() for now and potentially check in the future if it makes sense to switch (we don't necessarily need to hold up this PR).

Reply:
We can yeah, good call. I think the API is the same.

Reply:
Turns out it's not as straightforward. Will make an issue. I doubt it makes a huge difference though.

Reply:
Ok no problem. We can use write.table().